src/java/org/apache/cassandra/db/filter/ColumnFilter.java

```diff
@@ -1043,28 +1072,50 @@ public ColumnFilter deserialize(DataInputPlus in, int version, TableMetadata met
                 // is assumed that all columns are queried.
                 if (!hasQueried || isUpgradingFromVersionLowerThan34())
                 {
-                    return new WildCardColumnFilter(fetched);
-                }
-
-                // pre CASSANDRA-12768 (4.0-) all static columns should be fetched along with all regular columns.
-                if (isUpgradingFromVersionLowerThan40())
-                {
-                    return new SelectionColumnFilter(FetchingStrategy.ALL_COLUMNS, queried, fetched, subSelections);
+                    return hasNoExtraStaticColumns && hasNoExtraRegularColumns ? new WildCardColumnFilter(fetched)
+                                                                               : new SelectionColumnFilter(FetchingStrategy.FETCHED_COLUMNS, fetched, fetched, subSelections);
```

Comment:
well, should it be ONLY_QUERIED_COLUMNS strategy in this case? although we are not fetching all the columns, all queried columns are fetched for sure